components/remote_settings/src/error.rs

```diff
@@ -0,0 +1,19 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#[derive(Debug, thiserror::Error)]
+pub enum RemoteSettingsError {
+    #[error("Network failure")]
+    NetworkFailure,
+    #[error("Error with HTTP client: {0}")]
+    ClientError(#[from] rs_client::ClientError),
+    #[error("JSON Error: {0}")]
+    JSONError(#[from] serde_json::Error),
+    #[error("Unexpected JSON format")]
+    InvalidJSONError,
```

Comment:
This is a side-effect of the potential error from deserializing with`serde_json::from_str`, but different errors when actually parsing as shown below. 

```
        let value: serde_json::Value = serde_json::from_str(&json_text)?;
        let data = value
            .get("data")
            .ok_or(RemoteSettingsError::InvalidJSONError)?;
        let vec = data
            .as_array()
            .ok_or(RemoteSettingsError::InvalidJSONError)?;
```

TBH, I would love to know a better way to do this kind of error comprehension in Rust, but it seems like each `Result` to be assigned separately when using the ? operator. For example, I would have written this in Kotlin more along the lines of:

```
fun parse_records(text: String): Result<List<JsonObject>> {
    Result.runCatching {
        // one liner to deserialize and parse because intermediate assignments aren't required
    }.getOrDefault(<custom error>)
}
```

Reply:
Aaah I see!

> seems like each Result to be assigned separately when using the ? operator.

Since it's all about the `From` implementation, we can have multiple errors map to our same variant! each type implements `From` itself, so if we are already returning a `RemoteSettingsError::InvalidJSONError`, it'll just map to the same variant, but if we have another error we need an explicit `From` implementation. 

`thiserror` gives us that `From` implementation for free when we do something like `JSONError(#[from] serde_json::Error)` (note the `#[from]`)

 if we do the suggestion in https://github.com/mozilla/application-services/pull/5423#discussion_r1128699873 we might get away with one error (i.e we'd remove the need to define our own `InvalidJSONError`


But if we don't, as long as `From<serde_json::Error>` returns the same variant as what our manual parsing returns, we should be in the clear!

we can do that by either:
- running `serde_json::from_str(&json_text).map_err(|_| RemoteSettingsError::InvalidJSONError)?` and removing the variant `JSONError` - that however means we lost context on the error string
- Changing `JSONError` to `JSONError(String)` and implementing our own, non-thiserror  `impl From<serde_json::Error> for RemoteSettingsError` that saves just the error string and not the error itself, then removing `InvalidJSONError` and replacing it with `JSONError(String)`